File: client/src/lib/game/combat.ts
import { Player, Enemy } from './entities';
import { useAudio } from '../stores/useAudio';

export function calculateDamage(attacker: Player | Enemy, defender: Player | Enemy): number {
  const baseDamage = attacker.attackPower;
  const defense = defender.defense;
  const damage = Math.max(1, baseDamage - defense);
  
  // Add some randomness (±25%)
  const variance = 0.25;
  const randomFactor = 1 + (Math.random() - 0.5) * variance * 2;
  
  return Math.floor(damage * randomFactor);
}

export function performAttack(attacker: Player | Enemy, defender: Player | Enemy): {
  damage: number;
  killed: boolean;
  message: string;
} {
  const damage = calculateDamage(attacker, defender);
  defender.health = Math.max(0, defender.health - damage);
  const killed = defender.health <= 0;
  
  // Play hit sound
  const audio = useAudio.getState();
  audio.playHit();
  
  const attackerName = 'level' in attacker ? 'You' : getEnemyName(attacker as Enemy);
  const defenderName = 'level' in defender ? 'You' : getEnemyName(defender as Enemy);
  
  let message = `${attackerName} attack${attackerName === 'You' ? '' : 's'} ${defenderName} for ${damage} damage!`;
  
  if (killed) {
    message += ` ${defenderName} ${defenderName === 'You' ? 'die' : 'dies'}!`;
  }
  
  return { damage, killed, message };
}

export function getEnemyName(enemy: Enemy): string {
  return enemy.type.charAt(0).toUpperCase() + enemy.type.slice(1);
}

export function gainExperience(player: Player, experience: number): {
  newPlayer: Player;
  leveledUp: boolean;
  message: string;
} {
  let newPlayer = { ...player };
  newPlayer.experience += experience;
  let leveledUp = false;
  let message = `You gain ${experience} experience!`;
  
  // Check for level up
  while (newPlayer.experience >= newPlayer.experienceToNext) {
    newPlayer.experience -= newPlayer.experienceToNext;
    newPlayer.level++;
    leveledUp = true;
    
    // Increase stats on level up
    const healthIncrease = Math.floor(Math.random() * 8) + 5; // 5-12 health
    const manaIncrease = Math.floor(Math.random() * 5) + 3;   // 3-7 mana
    const attackIncrease = Math.floor(Math.random() * 2) + 1; // 1-2 attack
    const defenseIncrease = Math.floor(Math.random() * 2) + 1; // 1-2 defense
    
    newPlayer.maxHealth += healthIncrease;
    newPlayer.health = newPlayer.maxHealth; // Full heal on level up
    newPlayer.maxMana += manaIncrease;
    newPlayer.mana = newPlayer.maxMana; // Full mana restore on level up
    newPlayer.attackPower += attackIncrease;
    newPlayer.defense += defenseIncrease;
    
    // Increase experience needed for next level
    newPlayer.experienceToNext = Math.floor(newPlayer.experienceToNext * 1.3);
    
    message += ` Level up! You are now level ${newPlayer.level}!`;
  }
  
  return { newPlayer, leveledUp, message };
}

File: python_game/game/combat.py
import random
from game.game_objects import Player, Enemy

def calculate_damage(attacker, defender):
    base_damage = attacker.attack_power
    defense = defender.defense
    damage = max(1, base_damage - defense)
    
    if not isinstance(attacker, Player):
        damage = damage * 3
    
    variance = 0.25
    random_factor = 1 + (random.random() - 0.5) * variance * 2
    
    return int(damage * random_factor)

def perform_attack(attacker, defender):
    damage = calculate_damage(attacker, defender)
    defender.health = max(0, defender.health - damage)
    killed = defender.health <= 0
    
    attacker_name = 'You' if isinstance(attacker, Player) else get_enemy_name(attacker)
    defender_name = 'You' if isinstance(defender, Player) else get_enemy_name(defender)
    
    message = f"{attacker_name} attack{'s' if attacker_name != 'You' else ''} {defender_name} for {damage} damage!"
    
    if killed:
        message += f" {defender_name} {'die' if defender_name == 'You' else 'dies'}!"
    
    return {'damage': damage, 'killed': killed, 'message': message}

def get_enemy_name(enemy: Enemy):
    return enemy.type.capitalize()

def gain_experience(player: Player, experience: int):
    player.experience += experience
    leveled_up = False
    message = f"You gain {experience} experience!"
    
    while player.experience >= player.experience_to_next:
        player.experience -= player.experience_to_next
        player.level += 1
        leveled_up = True
        
        health_increase = random.randint(5, 12)
        mana_increase = random.randint(3, 7)
        attack_increase = random.randint(1, 2)
        defense_increase = random.randint(1, 2)
        
        player.max_health += health_increase
        player.health = player.max_health
        player.max_mana += mana_increase
        player.mana = player.max_mana
        player.attack_power += attack_increase
        player.defense += defense_increase
        
        player.experience_to_next = int(player.experience_to_next * 1.3)
        
        message += f" Level up! You are now level {player.level}!"
    
    return {'leveled_up': leveled_up, 'message': message}
